core/src/main/scala/filodb.core/memstore/PartKeyLuceneIndex.scala

```diff
@@ -289,18 +298,58 @@ class PartKeyLuceneIndex(ref: DatasetRef,
     indexWriter.close()
   }
 
-  // TODO - most frequently used eviction needed
-  private val readerStateCache = mutable.WeakHashMap[(IndexReader, String), DefaultSortedSetDocValuesReaderState]()
+  // The reader state object is expensive to create, so reuse across queries.
+  // It is associated with reader/labelName
+  private val readerStateCache: LoadingCache[(IndexReader, String), DefaultSortedSetDocValuesReaderState] =
```

Comment:
Can you share the improvement we got by replacing ``WeakHashMap`` with ``LoadingCache``?

Reply:
I punted the caching problem earlier. WeakHashMap is not really a true cache. It works only until GC kicks in. The LoadingCache holds on to the state for longer. This will not really be measurable in our JMH benchmarks since they always query one label.